main.py

```diff
@@ -40,123 +43,144 @@ def __init__(self, settings: CommunitySettings) -> None:
+        self.pending_txs[tx.get_tx_hash()] = tx
+        # and this is correct that initially we send this tx only to receiver?
+        self.ez_send(receiver_peer, tx)
         self.counter += 1
-        print(f'[Node {self.get_peer_id(self.my_peer)}] Sending transaction {tx.nonce} to {self.get_peer_id(peer)}')
-        self.ez_send(peer, tx)
 
         # WIP: We need this?
         # if self.counter > self.max_messages:
         #     self.cancel_pending_task("tx_create")
         #     self.stop()
         #     return
 
-    def check_transactions(self):
-        print(f'[Node {self.get_peer_id(self.my_peer)}] Checking transactions')
+    def block_creation(self):
+        # WIP: get every known peer, not only the ones im connected to
+        peers = self.get_peers()
+        peers.append(self.my_peer)
+        selectedPeer = random.choice(peers)
+
+        if not selectedPeer.mid == self.my_peer.mid:
+            return
 
-        for tx in self.pending_txs:
-            if self.balances[tx.sender] - tx.amount >= 0:
-                self.balances[tx.sender] -= tx.amount
-                self.balances[tx.receiver] += tx.amount
-                self.pending_txs.remove(tx)
-                self.finalized_txs.append(tx)
-                self.current_block.add_transaction(tx)
+        logging.info(f'[Node {self.get_peer_id(self.my_peer)}] is creating a block')
+        for tx_hash in list(self.pending_txs.keys()):
+            tx = self.pending_txs.pop(tx_hash)
+            self.finalized_txs[tx_hash] = tx
+            self.current_block.add_transaction(tx)
 
-                if self.current_block.is_full():
-                    print('Block is full')
-                    self.finalize_and_broadcast_block()
-                    break
+            if self.current_block.is_full():
+                logging.info(f'[Node {self.get_peer_id(self.my_peer)}] Chosen one')
+                self.finalize_and_broadcast_block()
+                break
 
     @lazy_wrapper(Transaction)
-    async def on_transaction(self, peer: Peer, payload: Transaction) -> None:
+    async def on_transaction(self, peer: Peer, tx: Transaction) -> None:
         my_id = self.get_peer_id(self.my_peer)
+        logging.info(f'[Node {my_id}] received transaction {tx.nonce} from {self.get_peer_id(peer)}')
 
-        print(f'[Node {my_id}] Received transaction', payload.nonce, 'from', self.get_peer_id(peer))
-
-        # Add to pending transactions
-        if (payload.sender, payload.nonce) not in [(tx.sender, tx.nonce) for tx in self.finalized_txs] and (
-                payload.sender, payload.nonce) not in [(tx.sender, tx.nonce) for tx in self.pending_txs]:
-            self.pending_txs.append(payload)
+        tx_hash = tx.get_tx_hash()
+        # if we already have this tx we do nothing
+        # here if we have the same txs received from different peers - we choose the one that was sent earlier?
```

Comment:
I think both would have the same information right? so maybe we don't care about is the oldest, if we already have It we return